ipaserver/plugins/certmapping.py

```diff
@@ -239,11 +248,186 @@ class cert_get_requestdata(Command):
     )
 
     def execute(self, **kw):
-        principal = kw.get('principal')
+        principal_string = kw.get('principal')
         profile_id = kw.get('profile_id', self.Backend.ra.DEFAULT_PROFILE)
         helper = kw.get('format')
 
-        result = {'debug_output': u'test'}
+        # TODO(blipton): factor this out of here and cert_request()
+        principal = split_any_principal(principal_string)
+        servicename, principal_name, realm = principal
+
+        if servicename is None:
+            principal_type = USER
+        elif servicename == 'host':
+            principal_type = HOST
+        else:
+            principal_type = SERVICE
+
+        try:
+            if principal_type == SERVICE:
+                principal_obj = api.Command.service_show(principal_string, all=True)
+            elif principal_type == HOST:
+                principal_obj = api.Command.host_show(principal_name, all=True)
+            elif principal_type == USER:
+                principal_obj = api.Command.user_show(principal_name, all=True)
+        except errors.NotFound:
+            raise errors.NotFound(
+                reason=_("The principal for this request doesn't exist."))
+        principal_obj = principal_obj['result']
+
+        request_data = self.Backend.certmapping.get_request_data(principal_obj,
+                profile_id, helper)
+
+        result = {}
+        result['debug_output'] = unicode(request_data)
         return dict(
             result=result
         )
+
+class OpenSSLHelper(object):
+    # TODO(blipton): What if we want a parenthesis in the section?
+    SECTION_RE = re.compile(r'%section\((.*?)\)', re.DOTALL)
+    EXTENSION_RE = re.compile(r'^ext:(.*)$', re.DOTALL+re.MULTILINE)
+
+    def parse_sections(self, value, sections):
+        def process_section(match):
+            sec_name = 'sec%s' % len(sections)
+            sections.append(u'[%s]\n%s' % (sec_name, match.group(1)))
+            return sec_name
+
+        line = self.SECTION_RE.sub(process_section, value)
+        return line
+
+    def format(self, values):
+        all_lines = [u'[req]', u'prompt = no']
+        all_sections = []
+        all_extensions = []
+        for value in values:
+            match = self.EXTENSION_RE.match(value)
+            if match:
+                value = match.group(1)
+
+            line = self.parse_sections(value, all_sections)
+
+            if match:
+                all_extensions.append(line)
+            else:
+                all_lines.append(line)
+
+        if all_extensions:
+            all_lines.append(u'req_extensions = exts')
+            exts = u'\n'.join([u'[exts]'] + all_extensions)
+            all_sections.append(exts)
+
+        main_section = u'\n'.join(all_lines)
+
+        config = u'\n\n'.join([main_section] + all_sections)
+
+        return config
+
+class CertutilHelper(object):
```

Comment:
If you want to have common interface there, you should use abstract class and abstract method (format) and inherit it in CerutilHelper and OpenSSLHelper